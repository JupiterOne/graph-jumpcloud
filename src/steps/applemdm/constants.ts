import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { OrgEntities } from '../orgs/constants';

export const AppleEntities: Record<
  'APPLE_MDM' | 'APPLE_DEVICE',
  StepEntityMetadata
> = {
  APPLE_MDM: {
    _type: 'jumpcloud_apple_mdm',
    _class: ['Configuration'],
    resourceName: 'Configuration',
  },
  APPLE_DEVICE: {
    _type: 'jumpcloud_apple_device',
    _class: ['Device'],
    resourceName: 'Device',
  },
};
export const AppleRelationships: Record<
  'ORG_HAS_MDM_APPLE' | 'APPLE_DEVICE_HAS_MDM',
  StepRelationshipMetadata
> = {
  ORG_HAS_MDM_APPLE: {
    _type: 'jumpcloud_account_has_apple_mdm',
    _class: RelationshipClass.HAS,
    sourceType: OrgEntities.ORG._type,
    targetType: AppleEntities.APPLE_MDM._type,
  },
  APPLE_DEVICE_HAS_MDM: {
    _type: 'jumpcloud_apple_device_has_mdm',
    _class: RelationshipClass.HAS,
    sourceType: AppleEntities.APPLE_DEVICE._type,
    targetType: AppleEntities.APPLE_MDM._type,
  },
};
