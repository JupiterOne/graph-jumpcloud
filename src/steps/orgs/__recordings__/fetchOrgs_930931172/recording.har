{
  "log": {
    "_recordingName": "fetchOrgs",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "68692ed96e085eab550b21a15bbb26da",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-org-id",
              "value": ""
            },
            {
              "_fromType": "array",
              "name": "x-api-key",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "contenttype",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "console.jumpcloud.com"
            }
          ],
          "headersSize": 320,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://console.jumpcloud.com/api/organizations"
        },
        "response": {
          "bodySize": 288,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 288,
            "text": "{\"totalCount\":1,\"results\":[{\"_id\":\"60259a2ad81e1a474d4e7c83\",\"id\":\"60259a2ad81e1a474d4e7c83\",\"displayName\":null,\"logoUrl\":null,\"created\":\"2021-02-11T20:57:14.837Z\"}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "nginx"
            },
            {
              "name": "date",
              "value": "Fri, 12 Feb 2021 20:13:30 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "etag",
              "value": "W/\"a6-li2T9RYtoKUz1R9KdaIOzhovpiE\""
            },
            {
              "name": "x-response-time",
              "value": "28.441ms"
            },
            {
              "name": "content-security-policy",
              "value": "child-src https://fast.wistia.net https://jumpcloud-1.wistia.com https://www.google.com https://youtube.com; connect-src 'self' https://api.segment.io https://app.launchdarkly.com https://cdn3.optimizely.com https://clientstream.launchdarkly.com https://device-cert.jumpcloud.com https://errors.client.optimizely.com https://events.launchdarkly.com https://ipv4.icanhazip.com https://logx.optimizely.com https://segment.jumpcloud.com https://solupay.transactiongateway.com https://www.google-analytics.com https://www.google.com; default-src 'self'; font-src 'self' data:; frame-ancestors https://fast.wistia.net https://jumpcloud-1.wistia.com https://www.google.com https://youtube.com; frame-src https://*.doubleclick.net https://*.duosecurity.com https://a553480116.cdn.optimizely.com/ https://fast.wistia.net https://js.driftt.com https://js.stripe.com https://jumpcloud-1.wistia.com https://solupay.transactiongateway.com https://www.google.com https://www.youtube.com https://youtube.com; img-src 'self' data: https://*.google.com https://ad.doubleclick.net https://alb.reddit.com https://assets.jumpcloud.com https://bat.bing.com https://chart.googleapis.com https://facebook.com https://googleads.g.doubleclick.net https://insight.adsrvr.org https://match.adsrvr.org https://p.typekit.net https://ping.typekit.net https://px.ads.linkedin.com https://q.stripe.com https://s3.amazonaws.com/static.jumpcloud.com/ https://stats.g.doubleclick.net https://storage.googleapis.com https://www.facebook.com https://www.google-analytics.com https://www.redditstatic.com; media-src https://fast.wistia.net https://js.driftt.com; object-src 'none'; script-src 'self' 'unsafe-eval' 'unsafe-inline' https://api.duosecurity.com https://api.stripe.com https://bat.bing.com https://cdn3.optimizely.com https://connect.facebook.com https://connect.facebook.net https://ct.capterra.com https://fast.wistia.net https://googleads.g.doubleclick.net https://js.driftt.com https://js.stripe.com https://pi.pardot.com https://px.ads.linkedin.com https://q.stripe.com https://rum.optimizely.com https://sjs.bizographics.com https://solupay.transactiongateway.com https://stats.g.doubleclick.net https://storage.googleapis.com https://tagmanager.google.com/ https://thirdparty.jumpcloud.com https://use.typekit.net https://www.google-analytics.com https://www.google.com https://www.googleadservices.com https://www.googletagmanager.com/ https://www.gstatic.com https://www.redditstatic.com; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com/ https://solupay.transactiongateway.com https://tagmanager.google.com/ https://use.typekit.net"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains;"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "x-ua-compatible",
              "value": "IE=edge"
            },
            {
              "name": "cache-control",
              "value": "no-store"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            }
          ],
          "headersSize": 3126,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-02-12T20:13:30.163Z",
        "time": 197,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 197
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
