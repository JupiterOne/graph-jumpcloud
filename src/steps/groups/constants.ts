import { RelationshipClass, StepEntityMetadata, StepRelationshipMetadata } from "@jupiterone/integration-sdk-core";
import { OrgEntities } from "../orgs/constants";
import { UserEntities } from "../users/constants";

export const GroupEntities: Record<
  'GROUP',
  StepEntityMetadata
> = {
  GROUP: {
    _type: 'jumpcloud_group',
    _class: ['Group'],
    resourceName: 'Group',
  },
};

export const GroupRelationships: Record<
  'ORG_HAS_GROUP' |
  'GROUP_HAS_USER',
  StepRelationshipMetadata
> = {
  ORG_HAS_GROUP: {
    _type: 'jumpcloud_account_has_group',
    _class: RelationshipClass.HAS,
    sourceType: OrgEntities.ORG._type,
    targetType: GroupEntities.GROUP._type,
  },
  GROUP_HAS_USER: {
    _type: 'jumpcloud_group_has_user',
    _class: RelationshipClass.HAS,
    sourceType: GroupEntities.GROUP._type,
    targetType: UserEntities.USER._type,
  },
};
