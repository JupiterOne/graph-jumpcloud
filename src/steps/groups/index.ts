import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { JumpCloudClient } from '../../jumpcloud/client';
import { IntegrationConfig } from '../../config';
import { GroupEntities, GroupRelationships} from './constants';
import { createGroupEntity } from './converters';
import { PRIORITY_ORG_CACHE_KEY } from '../orgs/constants';

export async function fetchGroups({
  instance,
  jobState,
  logger
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const client = new JumpCloudClient({
    logger,
    apiKey: instance.config.apiKey,
    orgId: instance.config.orgId
  });

  const orgEntity = await jobState.getData<Entity>(PRIORITY_ORG_CACHE_KEY);

  await client.iterateGroups(async (group) => {
    const groupEntity = await jobState.addEntity(
      createGroupEntity(group)
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: orgEntity,
        to: groupEntity,
      }),
    );
  });
}

export async function fetchGroupMembers({
  instance,
  jobState,
  logger
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const client = new JumpCloudClient({
    logger,
    apiKey: instance.config.apiKey,
    orgId: instance.config.orgId
  });

  await jobState.iterateEntities({
    _type: GroupEntities.GROUP._type
  }, async (groupEntity) => {
    const groupId = groupEntity.id as string;

    await client.iterateGroupMembers(
      groupId,
      async (member) => {
        const memberUserId = member.to?.id;

        if (!memberUserId) {
          return;
        }

        const userEntity = await jobState.findEntity(memberUserId);

        if (!userEntity) {
          // It's possible that this user was created in between the time that
          // we collected all users and now. We should just ignore this one and
          // carry on!
          return
        }

        await jobState.addRelationship(
          createDirectRelationship({
            _class: RelationshipClass.HAS,
            from: groupEntity,
            to: userEntity,
          }),
        );
      }
    );
  });
}

export const groupSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-groups',
    name: 'Fetch Groups',
    entities: [GroupEntities.GROUP],
    relationships: [
      GroupRelationships.ORG_HAS_GROUP,
    ],
    dependsOn: ['fetch-orgs'],
    executionHandler: fetchGroups,
  },
  {
    id: 'fetch-group-members',
    name: 'Fetch Group Members',
    entities: [],
    relationships: [
      GroupRelationships.GROUP_HAS_USER,
    ],
    dependsOn: ['fetch-groups'],
    executionHandler: fetchGroupMembers,
  },
];
