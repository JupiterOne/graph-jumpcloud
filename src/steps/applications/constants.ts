import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { GroupEntities } from '../groups/constants';
import { UserEntities } from '../users/constants';
import { OrgEntities } from '../orgs/constants';

export const ApplicationEntities: Record<'APPLICATION', StepEntityMetadata> = {
  APPLICATION: {
    _type: 'jumpcloud_application',
    _class: ['Application'],
    resourceName: 'Application',
  },
};

export const ApplicationRelationships: Record<
  | 'USER_ASSIGNED_APPLICATION'
  | 'GROUP_ASSIGNED_APPLICATION'
  | 'ORG_HAS_APPLICATION',
  StepRelationshipMetadata
> = {
  USER_ASSIGNED_APPLICATION: {
    _type: 'jumpcloud_user_assigned_application',
    _class: RelationshipClass.ASSIGNED,
    sourceType: UserEntities.USER._type,
    targetType: ApplicationEntities.APPLICATION._type,
  },
  GROUP_ASSIGNED_APPLICATION: {
    _type: 'jumpcloud_group_assigned_application',
    _class: RelationshipClass.ASSIGNED,
    sourceType: GroupEntities.GROUP._type,
    targetType: ApplicationEntities.APPLICATION._type,
  },
  ORG_HAS_APPLICATION: {
    _type: 'jumpcloud_account_has_application',
    _class: RelationshipClass.HAS,
    sourceType: OrgEntities.ORG._type,
    targetType: ApplicationEntities.APPLICATION._type,
  },
};
