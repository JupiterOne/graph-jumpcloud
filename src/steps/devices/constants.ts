import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { ApplicationEntities } from '../applications/constants';
import { UserEntities } from '../users/constants';

export const DeviceEntities: Record<'DEVICE', StepEntityMetadata> = {
  DEVICE: {
    _type: 'jumpcloud_device',
    _class: ['Device'],
    resourceName: 'Device',
  },
};

export const DeviceRelationships: Record<
  'DEVICE_INSTALLED_APPLICATION' | 'USER_HAS_DEVICE',
  StepRelationshipMetadata
> = {
  DEVICE_INSTALLED_APPLICATION: {
    _type: 'jumpcloud_device_installed_application',
    _class: RelationshipClass.INSTALLED,
    sourceType: DeviceEntities.DEVICE._type,
    targetType: ApplicationEntities.APPLICATION._type,
  },
  USER_HAS_DEVICE: {
    _type: 'jumpcloud_user_has_device',
    _class: RelationshipClass.HAS,
    sourceType: UserEntities.USER._type,
    targetType: DeviceEntities.DEVICE._type,
  },
};
