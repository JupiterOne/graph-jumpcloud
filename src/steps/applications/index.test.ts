import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import { fetchApplicationMembers, fetchApplications } from '.';
import { withRecording } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { IntegrationConfig } from '../../config';
import { ApplicationEntities, ApplicationRelationships } from './constants';
import { fetchOrgs } from '../orgs';
import { RelationshipClass } from '@jupiterone/integration-sdk-core';
import { fetchUsers } from '../users';
import { fetchGroups } from '../groups';

describe('#fetchApplications', () => {
  test('should collect data', async () => {
    await withRecording('fetchApplications', __dirname, async () => {
      const context = createMockStepExecutionContext<IntegrationConfig>({
        instanceConfig: integrationConfig,
      });

      await fetchOrgs(context);
      await fetchApplications(context);

      expect({
        numCollectedEntities: context.jobState.collectedEntities.length,
        numCollectedRelationships:
          context.jobState.collectedRelationships.length,
        collectedEntities: context.jobState.collectedEntities,
        collectedRelationships: context.jobState.collectedRelationships,
        encounteredTypes: context.jobState.encounteredTypes,
      }).toMatchSnapshot();

      expect(
        context.jobState.collectedEntities.filter(
          (e) => e._type === ApplicationEntities.APPLICATION._type,
        ),
      ).toMatchGraphObjectSchema({
        _class: ['Application'],
        schema: {
          additionalProperties: true,
          properties: {
            _type: { const: 'jumpcloud_application' },
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            attributes: { type: 'string' },
          },
        },
      });

      expect(
        context.jobState.collectedRelationships.filter(
          (r) => r._type === ApplicationRelationships.ORG_HAS_APPLICATION._type,
        ),
      ).toMatchDirectRelationshipSchema({
        schema: {
          properties: {
            _class: { const: RelationshipClass.HAS },
            _type: { const: 'jumpcloud_account_has_application' },
          },
        },
      });
    });
  });
});

describe('#fetchAppBoundUsersAndGroups', () => {
  test('should collect data', async () => {
    await withRecording('fetchApplications', __dirname, async () => {
      const context = createMockStepExecutionContext<IntegrationConfig>({
        instanceConfig: integrationConfig,
      });

      await fetchOrgs(context);
      await fetchUsers(context);
      await fetchGroups(context);
      await fetchApplications(context);
      await fetchApplicationMembers(context);

      expect({
        numCollectedEntities: context.jobState.collectedEntities.length,
        numCollectedRelationships:
          context.jobState.collectedRelationships.length,
        collectedEntities: context.jobState.collectedEntities,
        collectedRelationships: context.jobState.collectedRelationships,
        encounteredTypes: context.jobState.encounteredTypes,
      }).toMatchSnapshot();

      expect(
        context.jobState.collectedRelationships.filter(
          (r) =>
            r._type ===
            ApplicationRelationships.USER_ASSIGNED_APPLICATION._type,
        ),
      ).toMatchDirectRelationshipSchema({
        schema: {
          properties: {
            _class: { const: RelationshipClass.ASSIGNED },
            _type: { const: 'jumpcloud_user_assigned_application' },
          },
        },
      });

      expect(
        context.jobState.collectedRelationships.filter(
          (r) =>
            r._type ===
            ApplicationRelationships.GROUP_ASSIGNED_APPLICATION._type,
        ),
      ).toMatchDirectRelationshipSchema({
        schema: {
          properties: {
            _class: { const: RelationshipClass.ASSIGNED },
            _type: { const: 'jumpcloud_group_assigned_application' },
          },
        },
      });
    });
  });
});
