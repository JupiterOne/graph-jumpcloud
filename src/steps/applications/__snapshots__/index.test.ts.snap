// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchAppBoundUsersAndGroups should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Account",
        "Organization",
      ],
      "_key": "60259a2ad81e1a474d4e7c83",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "_id": "60259a2ad81e1a474d4e7c83",
            "created": "2021-02-11T20:57:14.837Z",
            "displayName": null,
            "id": "60259a2ad81e1a474d4e7c83",
            "logoUrl": null,
          },
        },
      ],
      "_type": "jumpcloud_account",
      "createdOn": 1613077034837,
      "displayName": "60259a2ad81e1a474d4e7c83",
      "id": "60259a2ad81e1a474d4e7c83",
      "logoUrl": undefined,
      "name": "60259a2ad81e1a474d4e7c83",
      "webLink": "https://console.jumpcloud.com",
    },
    {
      "_class": [
        "User",
      ],
      "_key": "6026bec6b8207255280d5981",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "_id": "6026bec6b8207255280d5981",
            "account_locked": false,
            "account_locked_date": null,
            "activated": false,
            "addresses": [],
            "allow_public_key": true,
            "attributes": [],
            "company": "",
            "costCenter": "",
            "created": "2021-02-12T17:45:42.369Z",
            "department": "",
            "description": "Testing desc",
            "disableDeviceMaxLoginAttempts": false,
            "displayname": "Austin K",
            "email": "austin.kelleher+testjc@jupiterone.com",
            "employeeIdentifier": null,
            "employeeType": "",
            "enable_managed_uid": false,
            "enable_user_portal_multifactor": false,
            "external_dn": "",
            "external_source_type": "",
            "externally_managed": false,
            "firstname": "Austin",
            "id": "6026bec6b8207255280d5981",
            "jobTitle": "",
            "lastname": "Kelleher",
            "ldap_binding_user": false,
            "location": "",
            "mfa": {
              "configured": false,
              "exclusion": false,
            },
            "middlename": "",
            "password_expired": false,
            "password_never_expires": false,
            "passwordless_sudo": false,
            "phoneNumbers": [],
            "samba_service_user": false,
            "ssh_keys": [],
            "state": "ACTIVATED",
            "sudo": false,
            "suspended": false,
            "systemUsername": "",
            "tags": [],
            "totp_enabled": false,
            "unix_guid": 5001,
            "unix_uid": 5001,
            "userTags": undefined,
            "username": "austinkellehertest",
          },
        },
      ],
      "_type": "jumpcloud_user",
      "active": false,
      "company": "",
      "createdOn": 1613151942369,
      "department": "",
      "description": "Testing desc",
      "displayName": "austinkellehertest",
      "email": "austin.kelleher+testjc@jupiterone.com",
      "employeeId": undefined,
      "employeeType": "",
      "externallyManaged": false,
      "firstName": "Austin",
      "id": "6026bec6b8207255280d5981",
      "jobTitle": "",
      "lastName": "Kelleher",
      "ldapBindingUser": false,
      "login": "austinkellehertest",
      "mfaConfigured": false,
      "mfaEnabled": false,
      "mfaExclusion": false,
      "mfaExclusionUntil": undefined,
      "name": "Austin Kelleher",
      "sambaServiceUser": false,
      "sudo": false,
      "suspended": false,
      "username": "austinkellehertest",
      "webLink": "https://console.jumpcloud.com/#/users/6026bec6b8207255280d5981/details",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "60259a3cb544065d1eaf2760",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "attributes": null,
            "description": "",
            "email": "",
            "id": "60259a3cb544065d1eaf2760",
            "name": "All Users",
            "type": "user_group",
          },
        },
      ],
      "_type": "jumpcloud_group",
      "attributes": undefined,
      "createdOn": undefined,
      "description": "",
      "displayName": "All Users",
      "id": "60259a3cb544065d1eaf2760",
      "name": "All Users",
      "webLink": "https://console.jumpcloud.com/#/groups/user/60259a3cb544065d1eaf2760/details",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "6026da081f24755306c2ee0d",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "attributes": null,
            "description": "Testing!",
            "email": "",
            "id": "6026da081f24755306c2ee0d",
            "name": "My Test Group",
            "type": "user_group",
          },
        },
      ],
      "_type": "jumpcloud_group",
      "attributes": undefined,
      "createdOn": undefined,
      "description": "Testing!",
      "displayName": "My Test Group",
      "id": "6026da081f24755306c2ee0d",
      "name": "My Test Group",
      "webLink": "https://console.jumpcloud.com/#/groups/user/6026da081f24755306c2ee0d/details",
    },
    {
      "_class": [
        "Application",
      ],
      "_key": "617c3664e6f8ae27f7e1201e",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "_id": "617c3664e6f8ae27f7e1201e",
            "active": true,
            "beta": false,
            "color": null,
            "created": "2021-10-29T17:59:00.448Z",
            "databaseAttributes": [],
            "description": "",
            "displayLabel": "AWS SSO",
            "displayName": "AWS SSO",
            "id": "617c3664e6f8ae27f7e1201e",
            "learnMore": "https://support.jumpcloud.com/support/s/article/Single-Sign-On-SSO-With-AWS-SSO",
            "logo": {
              "url": "https://s3.amazonaws.com/static.jumpcloud.com/applications/logos/application-icon-aws-sso.png",
            },
            "name": "aws-sso",
            "organization": "60259a2ad81e1a474d4e7c83",
            "sso": {
              "beta": false,
              "idpCertExpirationAt": "2026-10-29T17:58:59.000Z",
              "jit": false,
              "type": "saml",
            },
            "ssoUrl": "https://sso.jumpcloud.com/saml2/aws-sso",
          },
        },
      ],
      "_type": "jumpcloud_application",
      "active": true,
      "createdOn": 1635530340448,
      "description": "",
      "displayName": "AWS SSO",
      "id": "617c3664e6f8ae27f7e1201e",
      "name": "aws-sso",
      "ssoUrl": "https://sso.jumpcloud.com/saml2/aws-sso",
      "webLink": "https://console.jumpcloud.com/#/sso/617c3664e6f8ae27f7e1201e/details",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "HAS",
      "_fromEntityKey": "60259a2ad81e1a474d4e7c83",
      "_key": "60259a2ad81e1a474d4e7c83|has|6026bec6b8207255280d5981",
      "_toEntityKey": "6026bec6b8207255280d5981",
      "_type": "jumpcloud_account_has_user",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "60259a2ad81e1a474d4e7c83",
      "_key": "60259a2ad81e1a474d4e7c83|has|60259a3cb544065d1eaf2760",
      "_toEntityKey": "60259a3cb544065d1eaf2760",
      "_type": "jumpcloud_account_has_group",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "60259a2ad81e1a474d4e7c83",
      "_key": "60259a2ad81e1a474d4e7c83|has|6026da081f24755306c2ee0d",
      "_toEntityKey": "6026da081f24755306c2ee0d",
      "_type": "jumpcloud_account_has_group",
      "displayName": "HAS",
    },
    {
      "_class": "HAS",
      "_fromEntityKey": "60259a2ad81e1a474d4e7c83",
      "_key": "60259a2ad81e1a474d4e7c83|has|617c3664e6f8ae27f7e1201e",
      "_toEntityKey": "617c3664e6f8ae27f7e1201e",
      "_type": "jumpcloud_account_has_application",
      "displayName": "HAS",
    },
    {
      "_class": "ASSIGNED",
      "_fromEntityKey": "6026bec6b8207255280d5981",
      "_key": "6026bec6b8207255280d5981|assigned|617c3664e6f8ae27f7e1201e",
      "_toEntityKey": "617c3664e6f8ae27f7e1201e",
      "_type": "jumpcloud_user_assigned_application",
      "displayName": "ASSIGNED",
    },
    {
      "_class": "ASSIGNED",
      "_fromEntityKey": "6026da081f24755306c2ee0d",
      "_key": "6026da081f24755306c2ee0d|assigned|617c3664e6f8ae27f7e1201e",
      "_toEntityKey": "617c3664e6f8ae27f7e1201e",
      "_type": "jumpcloud_group_assigned_application",
      "displayName": "ASSIGNED",
    },
  ],
  "encounteredTypes": [
    "jumpcloud_account",
    "jumpcloud_user",
    "jumpcloud_account_has_user",
    "jumpcloud_group",
    "jumpcloud_account_has_group",
    "jumpcloud_application",
    "jumpcloud_account_has_application",
    "jumpcloud_user_assigned_application",
    "jumpcloud_group_assigned_application",
  ],
  "numCollectedEntities": 5,
  "numCollectedRelationships": 6,
}
`;

exports[`#fetchApplications should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Account",
        "Organization",
      ],
      "_key": "60259a2ad81e1a474d4e7c83",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "_id": "60259a2ad81e1a474d4e7c83",
            "created": "2021-02-11T20:57:14.837Z",
            "displayName": null,
            "id": "60259a2ad81e1a474d4e7c83",
            "logoUrl": null,
          },
        },
      ],
      "_type": "jumpcloud_account",
      "createdOn": 1613077034837,
      "displayName": "60259a2ad81e1a474d4e7c83",
      "id": "60259a2ad81e1a474d4e7c83",
      "logoUrl": undefined,
      "name": "60259a2ad81e1a474d4e7c83",
      "webLink": "https://console.jumpcloud.com",
    },
    {
      "_class": [
        "Application",
      ],
      "_key": "617c3664e6f8ae27f7e1201e",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "_id": "617c3664e6f8ae27f7e1201e",
            "active": true,
            "beta": false,
            "color": null,
            "created": "2021-10-29T17:59:00.448Z",
            "databaseAttributes": [],
            "description": "",
            "displayLabel": "AWS SSO",
            "displayName": "AWS SSO",
            "id": "617c3664e6f8ae27f7e1201e",
            "learnMore": "https://support.jumpcloud.com/support/s/article/Single-Sign-On-SSO-With-AWS-SSO",
            "logo": {
              "url": "https://s3.amazonaws.com/static.jumpcloud.com/applications/logos/application-icon-aws-sso.png",
            },
            "name": "aws-sso",
            "organization": "60259a2ad81e1a474d4e7c83",
            "sso": {
              "beta": false,
              "idpCertExpirationAt": "2026-10-29T17:58:59.000Z",
              "jit": false,
              "type": "saml",
            },
            "ssoUrl": "https://sso.jumpcloud.com/saml2/aws-sso",
          },
        },
      ],
      "_type": "jumpcloud_application",
      "active": true,
      "createdOn": 1635530340448,
      "description": "",
      "displayName": "AWS SSO",
      "id": "617c3664e6f8ae27f7e1201e",
      "name": "aws-sso",
      "ssoUrl": "https://sso.jumpcloud.com/saml2/aws-sso",
      "webLink": "https://console.jumpcloud.com/#/sso/617c3664e6f8ae27f7e1201e/details",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "HAS",
      "_fromEntityKey": "60259a2ad81e1a474d4e7c83",
      "_key": "60259a2ad81e1a474d4e7c83|has|617c3664e6f8ae27f7e1201e",
      "_toEntityKey": "617c3664e6f8ae27f7e1201e",
      "_type": "jumpcloud_account_has_application",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": [
    "jumpcloud_account",
    "jumpcloud_application",
    "jumpcloud_account_has_application",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;
